### PR TITLE
Add RaptorJIT

### PR DESCRIPTION
**Pull request by fsfod**
Add RaptorJIT to the VMs benchmarked.

### CODE REVIEW

File: build_luajit.sh [hunk 1]
```diff
@@ -17,6 +17,10 @@ else
     if [ ! -d "luajit_repo" ]; then 
       git clone -b v2.1 https://github.com/LuaJIT/LuaJIT luajit_repo
     fi
+    
+    if [ ! -d "ratorjit_repo" ]; then 
+      git clone https://github.com/raptorjit/raptorjit ratorjit_repo
```
Comment (vext01) [inline build_luajit.sh]: Should we be using a fixed tag/hash for our VMs?

File: luajit.krun [hunk 2]
```diff
@@ -46,6 +46,11 @@ VMS = {
         'variants': ['default-lua'],
         'n_iterations': ITERATIONS_ALL_VMS,
     },
+    'RaptorJIT': {
+        'vm_def': LuaVMDef("builds/raptorjit", env = VM_ENV),
```
Comment (vext01) [inline luajit.krun]: In Python, typically kwargs (keyword args) don't use spaces around `=`. I know, it's inconsistent, but that's the way the Python folk do it :P

File: luajit.krun [hunk 3]
```diff
@@ -69,6 +74,7 @@ BENCHMARKS = {
 # list of "bench:vm:variant"
 SKIP = [
     "capnproto_decode:GC64:*", # Sometime crashes with GC64
+    "capnproto_decode:RaptorJIT:*"
```
Comment (vext01) [inline luajit.krun]: Does this crash too?

File: build_luajit.sh (hunk 1 above)
Comment (ltratt) [reply build_luajit.sh]: Good point -- yes, we should.

Tom: what Edd means is that by fixing a hash, we know precisely which version of each JIT we're benchmarking.

That said, we could allow people to specific an option that means "just download the latest versions".

**Comment by vext01** [comment]
Travis choked!

File: luajit.krun (hunk 3 above)
Comment (fsfod) [reply luajit.krun]: It seems like it even though I'm not building it as GC64

Comment (ltratt) [reply luajit.krun]: What sort of crash are you seeing? We should probably report that upstream to raptorjit.

**Comment by fsfod** [comment]
He killed off the bootstrap support using a single Puc Lua file. so you have to have LuaJIT installed to run dynasm and stuff.

**Comment by vext01** [comment]
You can add packages inside travis if that helps.

File: luajit.krun (hunk 3 above)
Comment (vext01) [reply luajit.krun]: If it's crashing, let's add a comment saying that.

**Comment by fsfod** [comment]
That was what i was doing but i could just reuse the LuaJIT we already built.

**Comment by vext01** [comment]
That's also fine :)

File: build_luajit.sh [hunk 4]
```diff
@@ -26,6 +26,8 @@ fi
 ljbins=${ljbins:=builds}
 
 mkdir -p ${ljbins}
+# Make directory path relative so we don't have to worry about what path we have to pass to make.
+ljbins="$( cd "$ljbins" && pwd )"
```
Comment (vext01) [inline build_luajit.sh]: `${ljbins}` for consistnecy?

What does `$(...)` do here? Never seen this before.

File: build_luajit.sh (hunk 1 above)
Comment (fsfod) [reply build_luajit.sh]: I added locking to a commit hash

File: build_luajit.sh [hunk 5]
```diff
@@ -21,6 +20,13 @@ else
     if [ ! -d "raptorjit_repo" ]; then 
       git clone https://github.com/raptorjit/raptorjit raptorjit_repo
     fi
+
+    if [ -f "commitlock.txt" ]; then 
+      source commitlock.txt
+      echo "Current luajit commit hash: ${luajit_commitid}"
+      echo "Current raptorjit commit hash: ${raptorjit_commitid}"
+      (cd luajit_repo && git checkout ${luajit_commitid})
```
Comment (ltratt) [inline build_luajit.sh]: The `(...)` brackets aren't necessary.

File: build_luajit.sh (hunk 4 above)
Comment (fsfod) [reply build_luajit.sh]: I assume capture the stdout of the command run inside it.

File: build_luajit.sh (hunk 5 above)
Comment (fsfod) [reply build_luajit.sh]: I have to change the directory back if i don't run it like that

File: luajit.krun (hunk 2 above)
Comment (fsfod) [reply luajit.krun]: fixed

File: build_luajit.sh (hunk 5 above)
Comment (ltratt) [reply build_luajit.sh]: Oh, interesting -- I didn't realise it had that effect!

File: build_luajit.sh (hunk 4 above)
Comment (vext01) [reply build_luajit.sh]: Ah subshell. Fair enough.

I always use backticks, but this is fine too.

File: build_luajit.sh (hunk 5 above)
Comment (vext01) [reply build_luajit.sh]: today I learned!

**Comment by vext01** [comment]
What is the status of this @fsfod? Looks like you have addressed all of the comments?

**Comment by fsfod** [comment]
@ vext01 The only thing I didn't do was add the commitlock file or an opt out out for it but maybe that can be done when we pick a fixed target.

**Comment by vext01** [comment]
Up to you. I don't know the purpose of the commitlock file.

I'm OK to merge this as long as your next PR fixes the VM versions.

Do you want to squash?

**Comment by fsfod** [comment]
The commitlock will contain the commit hash ids to checkout when building LuaJIT and RaptorJIT from source with build_luajit.sh

**Comment by vext01** [comment]
OK, so that seems more like a change for the next PR then? What do you think?

**Comment by fsfod** [comment]
Yeah lets just squash.

**Comment by vext01** [comment]
Done?

**Comment by fsfod** [comment]
@vext01 yes.

**Comment by vext01** [comment]
OK. So next let's fix those VM versions.

**Pull request merged**